### ISSUE TITLE
Need explanation of IFunction syntax and how to use it

### ISSUE THREAD

**Issue by vgupta1127** [stoplightio/spectral#705]
**Chore summary**
Need more explanation of IFunction Syntax and how to use it

**Tasks**
Explain the different terms in IFunction syntax
How to use them

**Additional context**
[Custom Functions Docs Page](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/guides/custom-functions.md#custom-functions)
[GitHub Page for IFunction](https://github.com/stoplightio/spectral/blob/90a0864863fa232bf367a26dace61fd9f93198db/src/types/function.ts#L3#L8)

**Comment by P0lip** [comment]
@vgupta1127 
Hello!
`targetValue` - is the value you lint against, it's based on `given` jsonpath defined on the rule.
For instance, given the following document:
```yaml
openapi: 3.0.0
info:
  title: foo
```
and the following given path `$`, targetValue would be an object containing `openapi` and `info` properties.
If changed the path to `$.info.title`, `targetValue` would equal `foo`.

`options` is what a given custom function is provided with. Each rule can specify options that should be passed to the function. You can do it in the following way:
```yaml
operation-id-kebab-case:
    given: "$"
    then:
      function: pattern
      functionOptions:  # this will be exposed as options in the custom function
        match: ^[a-z][a-z0-9\-]*$
```

`paths` contains jsonpath expressions you specified in a rule.
Note, if a particular rule has a`field` property in `then`, that path will be exposed in as `target`.

`IFunctionValues` - the `original` and `given` are equal for the most of time. If you specify `field` I described above, then `original` and `given` will be an object that was matched using jsonpath expression provided in `given`. If `field` is not provided, they should be equal to `targetValue`.
Usage of `resolved` in `IFunctionValues` is discouraged.

Hope that helps.
Do let me know if you have any questions or if something is unclear.
More than happy to help and explain it further.

**Comment by vgupta1127** [comment]
Thanks Jakub for the explanation. But I'm still not clear. It would be great if you could show me a custom function that checks  the below problem statement...

Let's say I have the given API: [PetStore JSON](https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v2.0/json/petstore.json)

**I want to create a function that checks that GET operation object,(if present in the paths object) MUST HAVE a "200" resonse object and MUST NOT have a "201" response object.**

This task is more complex as it involves checking the operation(get) object's name and then checking all keys(for ex, 200/default/201...etc) in its "responses" object.

How would I create such a custom function and its associated rule?

**Comment by P0lip** [comment]
@vgupta1127 
If I understood the problem correctly, the following rule should do the trick:
- ruleset
```json
{
  "functions": ["validResponses"],
  "functionsDir": ".",
  "rules": {
    "get-operation-has-valid-responses": {
      "recommended": true,
      "given": "$.paths.*[?( @property === 'get' )].responses",
      "severity": "error",
      "message": "GET operation {{error}}",
      "then": {
        "function": "validResponses"
      }
    }
  }
}
```
- custom function called `validResponses.js`
````js
module.exports = (targetVal) => {
  if (typeof targetVal === 'object' && targetVal !== null) {
    if (!('200' in targetVal) && !('default' in targetVal)) {
      return [
        {
          message: 'must have either 200 or default response',
        },
      ];
    }

    if ('201' in targetVal) {
      return [
        {
          message: 'must not have both 200/default and 201 responses',
        },
      ];
    }
  }
};
```

Do let me know if it does the trick for you or if you have any other questions. More than happy to help. :)

**Comment by philsturgeon** [comment]
@P0lip maybe we could convert this support request into a request for improvements to the documentation? We could add the extra information supplied here to the docs, and add another more advanced example to the custom functions guide? Generally when folks have read the docs and still have questions it means we could improve our docs. :)

**Comment by vgupta1127** [comment]
@P0lip, Thanks for quick reply and a great explanation! I didn't know there is `@property` available in Spectral custom functions. I was trying to get hold of 'get' object. I have learnt a new thing, thanks to you! :) 

I have one quick question:

Is `{{error}}` in :

`"message": "GET operation {{error}}",`

just a placeholder for the message returned by the function `validResponses` ? I mean will the final message be `GET operation must have either 200 or default response` OR `GET operation must not have both 200/default and 201 responses`? 

And if yes, could I replace `{{error}}` with any other word `{{anotherWord}}` ?

**Comment by P0lip** [comment]
@vgupta1127 
I'm really glad it helped!

>  I didn't know before there is @property available in Spectral custom functions. 
Besides `@property`, there are also `@path`, `@parent` and `@parentProperty`. :)


> "message": "GET operation {{error}}"

Yes, that's right. `{{error}}` is a placeholder for the message returned by the function, therefore

>  I mean will the final message be GET operation must have either 200 or default response OR GET operation must not have both 200/default and 201 responses?

is right. It will be a dynamic response based on what the function return.
Obviously you might remove `{{error}}` and put something constant, but I believe having a dynamic response will provide a more meaningful feedback.

> And if yes, could I replace {{error}} with any other word {{anotherWord}} ?

Yep! We provide:
- `{{description}}` - the value of description you set on each rule,
- `{{property}}` - stands for the last segment of error path 
- `{{path}}`  - the whole error path
- `{{value}}` - the linted value

**Comment by vgupta1127** [comment]
Hi @P0lip 
How are you doing?

I put a comment in this thread, because I need some help with a new rule I am working on.

I am able to do quite easily using help from Spectral Documentation: to check if the value matches something for a given field name.


**What I would like is to do the opposite: to parse the all responses(200, etc etc) in all paths of OAS 2.0 and check that in response body:**

**if there is a field's value matches "some pattern", then the field name has to be "THIS".**

Consider this API OAS 2.0 file:
```
{
  "swagger": "2.0",
  "info": {
    "title": "Spectral",
    "version": "1.0",
    "contact": {
      "name": "abc",
      "email": "abc@gmail.com"
    }
  },
  "host": "abc.com",
  "schemes": [
    "http"
  ],
  "produces": [
    "application/json"
  ],
  "consumes": [
    "application/json"
  ],
  "paths": {
    "/getReports": {
      "get": {
        "summary": "Your GET endpoint",
        "tags": [],
        "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "type": "object",
              "properties": {
                "warnings": {
                  "$ref": "https://abc.com/models/Warnings.v1.yaml"
                }
              }
            }
          }
        },
        "operationId": "get-getReports"
      }
    },
    "/getReport": {
      "get": {
        "summary": "Your GET endpoint",
        "tags": [],
        "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "type": "object",
              "properties": {
                "contactRef": {
                  "$ref": "https://abc.com/models/Relationship.v1.yaml"
                },
                "contactRefs": {
                  "$ref": "https://abc.com/models/Relationships.v1.yaml"
                }
              }
            }
          }
        },
        "operationId": "get-getReport"
      },
      "post": {
        "summary": "",
        "operationId": "post-getReport",
        "responses": {
          "200": {
            "description": "OK",
            "schema": {
              "type": "object",
              "properties": {
                "abcRef": {
                  "$ref": "https://abc.com/models/Relationship.v1.yaml"
                },
                "abcRefs": {
                  "$ref": "https://abc.com/models/Relationships.v1.yaml"
                }
              }
            }
          }
        }
      }
    }
  },
  "definitions": {
    "XYZ": {
      "title": "XYZ",
      "type": "object",
      "properties": {
        "id": {
          "type": "string"
        },
        "XYZ_child": {
          "type": "object",
          "properties": {
            "id": {
              "type": "number"
            },
            "name": {
              "type": "string"
            },
            "address": {
              "type": "string"
            }
          }
        }
      }
    }
  },
  "securityDefinitions": {
    "API Key - 1": {
      "name": "API Key",
      "type": "apiKey",
      "in": "header"
    }
  }
}

```

What I would like is to go to properties to all response bodies of all paths and check that:

- if a field's value is `https://abc.com/models/Relationship.v1.yaml` then the field name must be suffixed with `Ref`

- if a field's value is `https://abc.com/models/Relationships.v1.yaml` then the field name must be suffixed with `Refs`


I do `$.paths.*.*.responses.*.*.properties` on the above API file which returns:

```
[
  {
    "warnings": {
      "$ref": "https://abc.com/models/Warnings.v1.yaml"
    }
  },
  {
    "contactRef": {
      "$ref": "https://abc.com/models/Relationship.v1.yaml"
    },
    "contactRefs": {
      "$ref": "https://abc.com/models/Relationships.v1.yaml"
    }
  },
  {
    "abcRef": {
      "$ref": "https://abc.com/models/Relationship.v1.yaml"
    },
    "abc": {
      "$ref": "https://abc.com/models/Relationships.v1.yaml"
    }
  }
]
```

I need to check that for every field's value having this Relationship(s) reference, field name should be suffixed Ref(s), for example `abc` in last object should be actually `abcRefs`.


Is there also a way to check all fields in an API file and apply this rule globally?
Thanks
Vikas

**Comment by nulltoken** [comment]
@vgupta1127 Hey! As this looks like a different question, could you please open a new issue?

**Comment by P0lip** [comment]
@nulltoken @vgupta1127 I'm about to provide an answer though, so perhaps no new issue will be needed in the end :sweat_smile: Hold on for a short while.

**Comment by P0lip** [comment]
@vgupta1127 could you please try the following rule?
```json
{
  "given": "$.paths.*.*.responses.*.*.properties",
  "then": {
    "function": "schema",
    "functionOptions": {
      "schema": {
        "additionalProperties": false,
        "patternProperties": {
          "Ref$": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
              "$ref": {
                "type": "string",
                "pattern": "[a-zA-Z][^s].v1.yaml$"
              }
            }
          },
          "Refs$": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
              "$ref": {
                "type": "string",
                "pattern": "[a-zA-Z]s.v1.yaml$"
              }
            }
          }
        }
      }
    }
  }
}
```

**Comment by P0lip** [comment]
Feel free to provide some meaningful `message`, so the error is slightly easier to understand.

**Comment by rossmcdonald** [comment]
@vgupta1127 Please direct these kinds of questions in the future to support@stoplight.io. That way we can make sure all questions get answered and nothing slips through the cracks.

**Comment by vgupta1127** [comment]
Sure, Thanks @rossmcdonald @P0lip.
